File: src/ios/Classes/Pipeline/PipelineRunner.swift
import AVFoundation
import Foundation

/// THE central object. Owns: pre-allocated frame, stage iteration,
/// deadline enforcement, error handling, and FramePublisher reference.
///
/// On iOS the AVAudioEngine tap callback IS the capture thread.
/// CapturePipeline provides the buffer handler that runs within it.
///
/// Uses double-buffering to eliminate the data race between the audio
/// thread (writes to the active frame) and the publisher queue (reads
/// from the inactive frame to serialize it).
final class PipelineRunner {
    private let config: PipelineConfig
    private let stages: [PipelineStage]
    private let publisher: FramePublisher
    private let framePeriodNanos: Int64
    private let deadlineNanos: Int64

    // Per-stage disabled flags — once a stage exceeds maxConsecutiveFailures
    // it is skipped for subsequent frames.
    private var stageDisabled: [Bool]
    private var stageFailures: [Int]

    // Double-buffer: the audio thread writes to the active frame while
    // the publisher queue reads from the inactive frame. They swap on
    // each publish so neither thread touches the same buffer simultaneously.
    private let frameA: MutableAudioFrame
    private let frameB: MutableAudioFrame
    private var activeFrame: MutableAudioFrame

    private static let maxConsecutiveFailures = 50

    init(config: PipelineConfig, stages: [PipelineStage], publisher: FramePublisher,
         frameSizeBytes: Int, sampleRate: Int = 16000, channelCount: Int = 1) {
        self.config = config
        self.stages = stages
        self.publisher = publisher
        self.framePeriodNanos = Int64(config.frameDurationMs) * 1_000_000
        self.deadlineNanos = Int64(Double(framePeriodNanos) * config.deadlineFraction)
        self.frameA = MutableAudioFrame(pcm: Data(count: frameSizeBytes))
        self.frameB = MutableAudioFrame(pcm: Data(count: frameSizeBytes))
        self.activeFrame = frameA
        self.stageDisabled = Array(repeating: false, count: stages.count)
        self.stageFailures = Array(repeating: 0, count: stages.count)
        self.frameA.sampleRate = sampleRate
        self.frameA.channelCount = channelCount
        self.frameB.sampleRate = sampleRate
        self.frameB.channelCount = channelCount
    }

    func initialize() {
        stages.forEach { $0.initialize(config: config) }
    }

    func start() {
        stages.forEach { $0.start() }
    }

    /// Called from the AVAudioEngine tap callback (internal audio thread).
    /// Copies buffer into the active frame, runs pipeline, then swaps
    /// buffers and publishes the previously-active frame.
    func handleBuffer(_ buffer: AVAudioPCMBuffer) {
        let byteCount = Int(buffer.frameLength) * MemoryLayout<Int16>.size * Int(buffer.format.channelCount)
        guard byteCount > 0, let channelData = buffer.int16ChannelData else { return }

        let frame = activeFrame
        frame.reset()
        if frame.pcm.count < byteCount {
            frame.pcm = Data(count: byteCount)
        }
        frame.pcm.withUnsafeMutableBytes { dest in
            memcpy(dest.baseAddress, channelData[0], byteCount)
        }
        frame.validBytes = byteCount

        let startNanos = DispatchTime.now().uptimeNanoseconds
        for (index, stage) in stages.enumerated() {
            if stageDisabled[index] { continue }
            do {
                stage.process(frame: frame)
                stageFailures[index] = 0
            } catch {
                stageFailures[index] += 1
                if stageFailures[index] >= Self.maxConsecutiveFailures && !stageDisabled[index] {
                    stageDisabled[index] = true
                    publisher.emitWarning(
                        code: "STAGE_DISABLED",
                        message: "Stage disabled after \(Self.maxConsecutiveFailures) consecutive failures"
                    )
                }
            }
        }

        frame.processingTimeNanos = Int64(DispatchTime.now().uptimeNanoseconds - startNanos)
        if frame.processingTimeNanos > deadlineNanos {
            publisher.emitWarning(code: "DEADLINE_EXCEEDED",
                                  message: "Frame \(frame.sequenceNumber) exceeded deadline")
        }

        // Swap buffers: the frame we just wrote becomes the publish
        // target, and the other frame becomes the active frame for the
        // next callback. The publisher queue reads the publish frame
        // while the audio thread writes to the new active frame.
        let publishFrame = frame
        activeFrame = (frame === frameA) ? frameB : frameA
        publisher.publish(frame: publishFrame)
    }

    func stop() {
        stages.forEach { $0.stop() }
    }

    func dispose() {
        stages.forEach { $0.dispose() }
    }
}

File: src/android/src/main/kotlin/com/circuids/boat/pipeline/PipelineRunner.kt
package com.circuids.boat.pipeline

import android.os.Process
import com.circuids.boat.audio.AudioCaptureEngine

/**
 * THE central object. Owns: capture thread, pre-allocated frame, stage iteration,
 * deadline enforcement, error handling, and FramePublisher reference.
 *
 * Capture is the producer (not a stage). Stages process frames after production.
 */
class PipelineRunner(
    private val config: PipelineConfig,
    private val stages: List<PipelineStage>,
    private val publisher: FramePublisher,
    private val captureEngine: AudioCaptureEngine,
    private val sampleRate: Int = 16000,
    private val channelCount: Int = 1,
) {
    private val framePeriodNanos = config.frameDurationMs * 1_000_000L
    private val deadlineNanos = (framePeriodNanos * config.deadlineFraction).toLong()
    private var consecutiveFailures = 0

    // Per-stage disabled flags — once a stage exceeds MAX_CONSECUTIVE_FAILURES
    // it is skipped for subsequent frames (not just warned about).
    private val stageDisabled = BooleanArray(stages.size)
    private val stageFailures = IntArray(stages.size)

    @Volatile
    private var running = false
    private var captureThread: Thread? = null

    @Volatile
    var captureFrameCount: Long = 0
        private set

    // Single pre-allocated frame, reused every iteration. Zero allocation per frame.
    private val frame = MutableAudioFrame(ByteArray(captureEngine.frameSizeBytes)).apply {
        this.sampleRate = this@PipelineRunner.sampleRate
        this.channelCount = this@PipelineRunner.channelCount
    }

    companion object {
        private const val MAX_CONSECUTIVE_FAILURES = 50
    }

    fun initialize() {
        stages.forEach { it.initialize(config) }
    }

    fun start() {
        if (running) return
        stages.forEach { it.start() }
        running = true
        captureEngine.startRecording()

        captureThread = Thread({
            Process.setThreadPriority(Process.THREAD_PRIORITY_URGENT_AUDIO)
            captureLoop()
        }, "boat-capture").also { it.start() }
    }

    /**
     * The capture loop: reset → read → stages → publish.
     * Runs on URGENT_AUDIO thread. No locks, no allocations, no thread hops.
     *
     * Wrapped in try/catch so an uncaught exception on the capture thread
     * emits an error event instead of dying silently.
     */
    private fun captureLoop() {
        try {
            while (running) {
                frame.reset()
                val bytesRead = captureEngine.read(frame.pcm, 0, frame.pcm.size)
                if (bytesRead <= 0) {
                    // Negative return (e.g. ERROR_DEAD_OBJECT) is fatal —
                    // break the loop and emit an error instead of spinning.
                    if (bytesRead < 0) {
                        publisher.emitError(
                            "CAPTURE_READ_FAILED",
                            "AudioRecord.read returned $bytesRead — capture device may be dead",
                        )
                        break
                    }
                    continue
                }
                frame.validBytes = bytesRead

                val startNanos = System.nanoTime()
                for ((index, stage) in stages.withIndex()) {
                    if (stageDisabled[index]) continue
                    try {
                        stage.process(frame)
                        stageFailures[index] = 0
                    } catch (_: Exception) {
                        stageFailures[index]++
                        if (stageFailures[index] >= MAX_CONSECUTIVE_FAILURES && !stageDisabled[index]) {
                            stageDisabled[index] = true
                            publisher.emitWarning(
                                "STAGE_DISABLED",
                                "Stage ${stage::class.simpleName} disabled after $MAX_CONSECUTIVE_FAILURES consecutive failures",
                            )
                        }
                    }
                }

                frame.processingTimeNanos = System.nanoTime() - startNanos
                if (frame.processingTimeNanos > deadlineNanos) {
                    publisher.emitWarning("DEADLINE_EXCEEDED", "Frame ${frame.sequenceNumber} exceeded deadline")
                }

                publisher.publish(frame)
                captureFrameCount++
            }
        } catch (e: Exception) {
            publisher.emitError("CAPTURE_THREAD_FATAL", "Capture thread terminated: ${e.message}")
        } finally {
            running = false
        }
    }

    fun stop() {
        if (!running) return
        running = false
        captureThread?.join(500)
        captureThread = null
        captureEngine.stop()
        stages.forEach { it.stop() }
    }

    fun dispose() {
        stop()
        stages.forEach { it.dispose() }
    }
}
